### ISSUE TITLE
batchnorm issue

### ISSUE THREAD

**Issue by xuchengggg** [dbolya/yolact#118]
Hi, I am very confused about the usage of batchnorm, in your code, whether I set the freeze_bn true or false, the eval results are same, but in my own code, I train the model with batchnorm, however I would get totally different results when I alter the training value in the batchnorm between true or false at the time of test. I look forward to your answer, thanks!

**Comment by dbolya** [comment]
Can you describe your process exactly? Setting freeze_bn to true or false should affect anything test time--only training time (and thus if you train with batch norm frozen you should get different results than if you train it without frozen, but if you took a model that's already been trained either way and just flip the setting for just evaluation, it shouldn't change anything).

If I'm understanding this correctly, you tried flipping the setting with the provided weights during eval and nothing changed (this is the correct behavior). Then you tried with "your code"? Does this mean your own implementation of YOLACT or that you tried training your own weights with my code?

**Comment by xuchengggg** [comment]
I implement YOLACT in tensorflow, and I use the
`class BatchNorm(keras.layers.BatchNormalization):
    def call(self, inputs, training=None):
        return super(self.__class__,self).call(inputs, training=training)`
as bn layer, in training time, I set the training=True,  but at test time I am not sure I should set the training true or false which would get different results.
By the way, what do you mean 
> Setting freeze_bn to true or false should affect anything test time--only training time .
Thank you for your reply!

**Comment by dbolya** [comment]
Oh that's completely different. This is not the place for debugging Keras code, since that's not a YOLACT issue >.> (you probably want the documentation for that layer instead), but whatever I can help you.

You don't want to set training=True during eval time because *you'd be overwriting the weights with every image you evaluate*. Freeze bn in my code *only affects training*. In Keras terms, it'd be like if I set that `training` variable in the constructor to: `training = False if freeze_bn else training` (so if freeze_bn is true, then set it to false no matter what, but if it isn't, then turn it on only when training).

That's what I mean that it should only affect anything while training: whether freeze_bn is on or not, that training should always be turned off during test time.

**Comment by xuchengggg** [comment]
Thank you very much!!
